airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py

```diff
@@ -0,0 +1,143 @@
+            yield record
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        last_modified = stream_state.get(self.cursor_field, self._start_datetime)
+        # since API filters inclusively, we add 1 sec to prevent duplicate reads
+        last_modified_dt = datetime.datetime.fromisoformat(last_modified)
+        last_modified_dt += datetime.timedelta(seconds=1)
+        last_modified = last_modified_dt.isoformat("T", "seconds")
+        return {"If-Modified-Since": last_modified}
+
+
+class ZohoStreamFactory:
+    def __init__(self, config: Mapping[str, Any]):
+        self.api = ZohoAPI(config)
+        self._config = config
+
+    def _init_modules_meta(self) -> List[ModuleMeta]:
+        modules_meta_json = self.api.modules_settings()
+        modules = [ModuleMeta.from_dict(module) for module in modules_meta_json]
+        return list(filter(lambda module: module.api_supported, modules))
+
+    def _populate_fields_meta(self, module: ModuleMeta):
+        fields_meta_json = self.api.fields_settings(module.api_name)
+        fields_meta = []
+        for field in fields_meta_json:
+            pick_list_values = field.get("pick_list_values", [])
+            if pick_list_values:
+                field["pick_list_values"] = [ZohoPickListItem.from_dict(pick_list_item) for pick_list_item in field["pick_list_values"]]
+            fields_meta.append(FieldMeta.from_dict(field))
+        module.fields = fields_meta
+
+    def _populate_module_meta(self, module: ModuleMeta):
+        module_meta_json = self.api.module_settings(module.api_name)
+        module.update_from_dict(next(iter(module_meta_json), None))
+
+    def produce(self) -> List[HttpStream]:
+        modules = self._init_modules_meta()
+        streams = []
+
+        def populate_module(module):
+            self._populate_module_meta(module)
+            self._populate_fields_meta(module)
+
+        def chunk(max_len, lst):
+            for i in range(math.ceil(len(lst) / max_len)):
+                yield lst[i * max_len : (i + 1) * max_len]
+
+        max_concurrent_request = self.api.max_concurrent_requests
+        with concurrent.futures.ThreadPoolExecutor(max_workers=max_concurrent_request) as executor:
+            for batch in chunk(max_concurrent_request, modules):
+                executor.map(lambda module: populate_module(module), batch)
+
+        for module in modules:
+            try:
+                schema = asdict(module.schema)
+            except (IncompleteMetaDataException, UnknownDataTypeException):
+                continue
+
+            stream_params = {
+                "url_base": f"{self.api.api_url}",
+                "_path": f"/crm/v2/{module.api_name}",
+                "json_schema": schema,
+                "primary_key": "id",
```

Comment:
If `primary_key` always static let’s move it in to `ZohoCrmStream` class

Reply:
done